### PR TITLE
GH #34: Support open() with barewords:

### PR DESCRIPTION
**Pull request by xsawyerx**
The last change to support barewords did not include `open()` by accident, but it wouldn't have helped anyway because of how `open()` operates with barewords calling them as `<FH>` and how `Test::MockFile` attempts to write over `$_[0]` which is what caused the original error expressed in the ticket (modification of read-only value).

So here's what we're doing differently:

* We added bareword upgrading to `open()`.

* We try to detect barewords using the READONLY flag which is what they will
  have.

* We retrieve the value of the bareword slot using the right caller, which
  is probably only relevant for barewords that already existed and have a
  value we want. I'm not sure what those situations will be, but okay.

* The bareword upgrading also now returns information on whether this is
  likely to be a bareword.

* This information is only used with `open()` at the moment to do two things:
  * Override `@_`, not `$_[0]`
  * Override the bareword slot using the caller information

The likely bareword information is removed in places where it's not used.

It's certainly more verbose but I think it's still for the better.

### CODE REVIEW

File: lib/Test/MockFile.pm [hunk 1]
```diff
@@ -1227,7 +1254,10 @@ BEGIN {
     };
 
     *CORE::GLOBAL::opendir = sub(*$) {
-        @_ = _upgrade_barewords(@_) if defined $_[0];
+        if ( defined $_[0] ) {
+            @_ = _upgrade_barewords(@_) if defined $_[0];
```
Comment (atoomic) [inline lib/Test/MockFile.pm:1258]: useless postfix check here `... if defined $_[0]` it was just checked above

File: lib/Test/MockFile.pm [hunk 2]
```diff
@@ -1227,7 +1254,10 @@ BEGIN {
...
+            shift;
```
Comment (atoomic) [inline lib/Test/MockFile.pm:1259]: maybe a comment or variable?
`shit; # drop likely_bareword flag`

File: lib/Test/MockFile.pm [hunk 3]
```diff
@@ -1274,7 +1304,10 @@ BEGIN {
     };
 
     *CORE::GLOBAL::readdir = sub(*) {
-        @_ = _upgrade_barewords(@_) if defined $_[0];
+        if ( defined $_[0] ) {
+            @_ = _upgrade_barewords(@_) if defined $_[0];
```
Comment (atoomic) [inline lib/Test/MockFile.pm:1308]: same comments here: extra useless postfix `defined` and comment for the `likely-bareword` flag?

File: lib/Test/MockFile.pm [hunk 4]
```diff
@@ -1312,7 +1345,10 @@ BEGIN {
     };
 
     *CORE::GLOBAL::telldir = sub(*) {
-        @_ = _upgrade_barewords(@_) if defined $_[0];
+        if ( defined $_[0] ) {
+            @_ = _upgrade_barewords(@_) if defined $_[0];
```
Comment (atoomic) [inline lib/Test/MockFile.pm:1349]: same

File: lib/Test/MockFile.pm [hunk 5]
```diff
@@ -1336,7 +1372,10 @@ BEGIN {
     };
 
     *CORE::GLOBAL::rewinddir = sub(*) {
-        @_ = _upgrade_barewords(@_) if defined $_[0];
+        if ( defined $_[0] ) {
+            @_ = _upgrade_barewords(@_) if defined $_[0];
```
Comment (atoomic) [inline lib/Test/MockFile.pm:1376]: same

File: lib/Test/MockFile.pm [hunk 6]
```diff
@@ -1361,7 +1400,10 @@ BEGIN {
     };
 
     *CORE::GLOBAL::seekdir = sub(*$) {
-        @_ = _upgrade_barewords(@_) if defined $_[0];
+        if ( defined $_[0] ) {
+            @_ = _upgrade_barewords(@_) if defined $_[0];
```
Comment (atoomic) [inline lib/Test/MockFile.pm:1404]: same

File: lib/Test/MockFile.pm [hunk 7]
```diff
@@ -1385,7 +1427,10 @@ BEGIN {
     };
 
     *CORE::GLOBAL::closedir = sub(*) {
-        @_ = _upgrade_barewords(@_) if defined $_[0];
+        if ( defined $_[0] ) {
+            @_ = _upgrade_barewords(@_) if defined $_[0];
```
Comment (atoomic) [inline lib/Test/MockFile.pm:1431]: same

**Review by atoomic — CHANGES_REQUESTED** [review]
LGTM please consider some minor cleanup as seen in comments

File: lib/Test/MockFile.pm (hunk 3 above)
Comment (toddr) [reply lib/Test/MockFile.pm:1308]: Is it useless on all versions of Perl?

File: lib/Test/MockFile.pm (hunk 1 above)
Comment (xsawyerx) [reply lib/Test/MockFile.pm:1258]: Yeah, that's a good point. I moved things around a lot and missed it. Just force-pushed a fixed commit.

File: lib/Test/MockFile.pm (hunk 2 above)
Comment (xsawyerx) [reply lib/Test/MockFile.pm]: Included the comment `# Remove bareword indicator`. Pushed.

File: lib/Test/MockFile.pm (hunk 3 above)
Comment (xsawyerx) [reply lib/Test/MockFile.pm]: Fixed.

File: lib/Test/MockFile.pm (hunk 4 above)
Comment (xsawyerx) [reply lib/Test/MockFile.pm]: Fixed.

File: lib/Test/MockFile.pm (hunk 5 above)
Comment (xsawyerx) [reply lib/Test/MockFile.pm]: Fixed.

File: lib/Test/MockFile.pm (hunk 6 above)
Comment (xsawyerx) [reply lib/Test/MockFile.pm]: Fixed.

File: lib/Test/MockFile.pm (hunk 7 above)
Comment (xsawyerx) [reply lib/Test/MockFile.pm]: Fixed.

**Comment by xsawyerx** [comment]
Force-pushed using `Symbol::qualify_to_ref` as a better method of retrieving the glob ref.

**Review by atoomic — APPROVED** [review]

File: lib/Test/MockFile.pm [hunk 8]
```diff
@@ -1227,7 +1254,10 @@ BEGIN {
     };
 
     *CORE::GLOBAL::opendir = sub(*$) {
-        @_ = _upgrade_barewords(@_) if defined $_[0];
+        if ( defined $_[0] ) {
+            @_ = _upgrade_barewords(@_);
+            shift; # Remove bareword indicator
```
Comment (atoomic) [inline lib/Test/MockFile.pm:1259]: LGTM, suggestion
```
( undef, @_ ) = upgrade_barewords(@_); # Remove bareword indicator
```

Comment (xsawyerx) [reply lib/Test/MockFile.pm:1259]: Good idea. I'll provide this in a separate PR.

**Pull request merged**